File: league/frontier.py
"""The frontier model, reached only through the gateway.

The OpenAI key lives in the Cloudflare gateway and nowhere else. The House asks
`POST <gateway>/v1/frontier/responses`; the gateway reserves the call's worst case against the
owner's monthly budget, forwards it, settles at what it really cost and says so in
`X-LTCM-Cost-USD`. A call that does not fit in what is left of the month is refused there (402),
not here, so nothing on Sail can spend past the line.
"""

from __future__ import annotations

import json
import re
import secrets
import urllib.error
import urllib.request
from dataclasses import dataclass
from decimal import Decimal, InvalidOperation
from typing import Any, Callable

MODEL = "gpt-6-astra"
COST_HEADER = "X-LTCM-Cost-USD"
AGENT_HEADER = "X-LTCM-Agent"
MAX_OUTPUT_TOKENS = 16000
# Verified standard-service ceilings, including long-context cache writes. Kept outside
# game.json: agents may choose a model, but cannot supply the price used to admit its bill.
# https://developers.openai.com/api/docs/pricing (2026-09-20), dollars / million tokens.
MODEL_CEILINGS = {
    "gpt-6-astra": (Decimal("25"), Decimal("75")),
    "gpt-5.6-sol": (Decimal("10"), Decimal("30")),
    "gpt-5.6-terra": (Decimal("5"), Decimal("18")),
    "gpt-5.6-luna": (Decimal("0.50"), Decimal("1.80")),
    # GPT-6 Sol and Luna, launched Sept 22, 2026 at half the GPT-5.6 rates: the long-context
    # cache-write input rate and the long-context output rate (gateway/wrangler.jsonc).
    "gpt-6-sol": (Decimal("5"), Decimal("15")),
    "gpt-6-luna": (Decimal("0.25"), Decimal("0.75")),
}


class FrontierError(RuntimeError):
    def __init__(self, message: str, *, status: int | None = None):
        super().__init__(message)
        self.status = status


@dataclass(frozen=True)
class Answer:
    text: str
    cost_usd: Decimal
    usage: dict[str, Any]
    model: str
    status: str = 'completed'
    cost_verified: bool = True

    def json(self) -> dict[str, Any]:
        """The first JSON object in the answer. Raises FrontierError when there is none."""
        if self.status != 'completed':
            raise FrontierError(f'the model response is {self.status}; no partial proposal is accepted')
        return extract_json(self.text)


def attribution(name: str) -> str:
    """The name the gateway files this call's cost under. It accepts `^[a-z0-9][a-z0-9_-]{0,63}$`
    and files anything else under "unattributed" (found on the night of the build: `audit:<agent>`
    was unattributed for its colon)."""
    clean = re.sub(r"[^a-z0-9_-]+", "-", str(name or "").lower()).strip("-_")[:64]
    return clean or "house"


def extract_json(text: str) -> dict[str, Any]:
    decoder = json.JSONDecoder()
    for index, char in enumerate(text):
        if char != "{":
            continue
        try:
            value, _ = decoder.raw_decode(text[index:])
        except ValueError:
            continue
        if isinstance(value, dict):
            return value
    raise FrontierError("the frontier model's answer held no JSON object")


def output_text(payload: dict[str, Any]) -> str:
    """The text of a Responses API result."""
    if isinstance(payload.get("output_text"), str) and payload["output_text"]:
        return payload["output_text"]
    parts = []
    for item in payload.get("output") or []:
        if isinstance(item, dict) and item.get("type") == "message":
            for block in item.get("content") or []:
                if isinstance(block, dict) and block.get("type") in ("output_text", "text") and isinstance(block.get("text"), str):
                    parts.append(block["text"])
    return "\n".join(parts)


class Frontier:
    def __init__(self, gateway_url: str, token_source: Callable[[], str], *, model: str = MODEL, opener: Any = None, timeout: float = 600.0, spend_guard: Any = None):
        self.url = gateway_url.rstrip("/") + "/v1/frontier/responses"
        self.token_source = token_source
        self.model = model
        self.opener = opener or urllib.request.urlopen
        self.timeout = timeout
        self.spend_guard = spend_guard

    def ask(self, *, system: str, user: str, agent: str, max_output_tokens: int = 6000, effort: str = "medium") -> Answer:
        return self.converse([{"role": "system", "content": system}, {"role": "user", "content": user}],
                             agent=agent, max_output_tokens=max_output_tokens, effort=effort)

    def converse(self, messages: list[dict[str, Any]], *, agent: str, max_output_tokens: int = 6000, effort: str = "medium",
                 cache: dict[str, Any] | None = None) -> Answer:
        """One call over a list of text messages, optionally with prompt-cache hints.

        `cache` carries `prompt_cache_key` / `prompt_cache_options` exactly as the Responses API
        names them; the gateway admits only bounded values (gateway/lib/frontier.mjs). A cached
        prefix changes the bill only through the usage block the gateway settles from."""
        body = {
            "model": self.model,
            "input": list(messages),
            "max_output_tokens": max(1, min(int(max_output_tokens), MAX_OUTPUT_TOKENS)),
            "reasoning": {"effort": effort},
        }
        for name in ("prompt_cache_key", "prompt_cache_options"):
            if cache and cache.get(name) is not None:
                body[name] = cache[name]
        request = urllib.request.Request(
            self.url, data=json.dumps(body).encode("utf-8"), method="POST",
            headers={"Authorization": "Bearer " + self.token_source(), "Content-Type": "application/json",
                     AGENT_HEADER: attribution(agent), "User-Agent": "ltcm-floor/1.0"},
        )
        commitment = "frontier:" + secrets.token_hex(16)
        if self.spend_guard is not None:
            if self.model not in MODEL_CEILINGS:
                raise FrontierError("the spend guard has no verified price for this model")
            # One UTF-8 byte per possible input token plus framing, including the long-context
            # and cache-write premiums. Standard service only; no built-in paid tools.
            input_rate, output_rate = MODEL_CEILINGS[self.model]
            hold = (Decimal(len(request.data) + 4096) * input_rate + Decimal(body["max_output_tokens"]) * output_rate) / 1000000
            try:
                self.spend_guard.reserve(commitment, "foundation-review", hold)
            except RuntimeError as exc:
                raise FrontierError(str(exc)) from None
        try:
            with self.opener(request, timeout=self.timeout) as response:
                payload = json.load(response)
                cost = response.headers.get(COST_HEADER)
        except urllib.error.HTTPError as exc:
            detail = exc.read()[:300].decode("utf-8", "replace")
            if self.spend_guard is not None and 400 <= exc.code < 500:
                # Refused, not billed: the gateway reserves nothing for its own refusals (402 month,
                # 400, 401, 403, 413) and settles an upstream 4xx at $0 (gateway/lib/router.mjs). Until
                # Sept 23, 2026 the House kept the worst-case hold of every refused call for good.
                self.spend_guard.settle(commitment, 0)
            raise FrontierError(f"frontier call refused: HTTP {exc.code} {detail}", status=exc.code) from None
        except (urllib.error.URLError, OSError, ValueError) as exc:
            raise FrontierError(f"frontier call failed: {type(exc).__name__}") from None
        if not isinstance(payload, dict):
            raise FrontierError("the frontier model's answer was not a JSON object")
        try:
            cost_usd = Decimal(str(cost)) if cost is not None else Decimal(0)
        except InvalidOperation:
            cost_usd = Decimal(0)
        if not cost_usd.is_finite() or cost_usd < 0:
            cost_usd = Decimal(0)  # the gateway's own meter is the record; a garbled header charges nothing here
        verified = False
        if cost is not None:
            try:
                confirmed = Decimal(str(cost))
                usage = payload.get('usage') if isinstance(payload.get('usage'), dict) else {}
                tokens_in, tokens_out = usage.get('input_tokens'), usage.get('output_tokens')
                if (confirmed.is_finite() and confirmed >= 0 and type(tokens_in) is int and type(tokens_out) is int
                        and min(tokens_in, tokens_out) >= 0 and payload.get('model') == self.model):
                    verified = True
                    # The gateway's metered cost, the same figure its month line enforces. It prices
                    # cache reads, cache writes and the long-context premiums from the provider's own
                    # usage block (gateway/lib/frontier.mjs actualCost, since Sept 22, 2026). Until
                    # Sept 23 the House booked max(this, every token at the long-context ceiling):
                    # measured then, the gateway's whole September was $267.29 while the House had
                    # settled $393.08 since Sept 21 alone, so the allowance closed at about half the
                    # owner's real spend. The hold is the ceiling; a verified call settles at its cost.
                    if self.spend_guard is not None:
                        self.spend_guard.settle(commitment, confirmed)
            except InvalidOperation:
                pass  # unknown costs retain the full hold, including across restarts
        return Answer(output_text(payload), cost_usd, dict(payload.get("usage") or {}), str(payload.get("model") or self.model),
                      str(payload.get('status') or 'completed'), verified)


class FrontierMonth:
    """What is left of the gateway's monthly frontier budget, read from `GET /v1/health`.

    The House's campaign meters its own holds, but the gateway's month is the line that really
    refuses (402), and it also carries holds the House never sees: a call that times out keeps its
    worst case there. On Sept 21, 2026 the campaign believed $77 remained while the gateway had $40,
    so cheap research would have spent the last of the month and left nothing for the audits that
    promotion to real money requires. `remaining()` is None when the gateway cannot be read; the
    callers then leave spending alone, and the gateway's own 402 is still the backstop.

    Since Sept 24, 2026 this month is also OpenAI's METER. Given the campaign (`meter`), every
    reading is fed to `CampaignBudget.observe_month`, which turns the month's spend into a figure
    that never falls, and the reader registers itself so that a reservation finding the last
    reading a minute old reads again first (`CampaignBudget.meter_reader`). The House reads it on
    every tick; a gateway that cannot be read leaves the meter unread, and after 180 s the campaign
    refuses OpenAI reservations -- paid OpenAI work stops, nothing else does."""

    def __init__(self, gateway_url: str, token_source: Callable[[], str], *, opener: Any = None, ttl: float = 60.0, clock: Any = None,
                 meter: Any = None):
        import threading
        import time as _time

        self.url = gateway_url.rstrip("/") + "/v1/health"
        self.token_source = token_source
        self.opener = opener or urllib.request.urlopen
        self.ttl = ttl
        self.clock = clock or _time.time
        self._at = float("-inf")
        self._value: Decimal | None = None
        self._bonus: Decimal | None = None
        self._lock = threading.Lock()  # one reading at a time: the tick and a reservation may both ask
        self.meter = meter
        if meter is not None:
            meter.meter_reader("openai", self.refresh)

    def remaining(self) -> Decimal | None:
        with self._lock:
            now = self.clock()
            if now - self._at < self.ttl:
                return self._value
            self._at = now
            reading = None
            try:
                request = urllib.request.Request(self.url, headers={"Authorization": "Bearer " + self.token_source(), "User-Agent": "ltcm-floor/1.0"})
                with self.opener(request, timeout=20) as response:
                    month = json.load(response).get("frontier") or {}
                value = Decimal(str(month["cap_usd"])) - Decimal(str(month["spent_usd"]))
                self._value = value if value.is_finite() else None
                self._bonus = None
                if month.get("base_cap_usd") is not None and self._value is not None:
                    # The cap in force less the configured month: what profit on the real accounts
                    # added (gateway/lib/equity.mjs). Absent from a gateway that does not index.
                    bonus = Decimal(str(month["cap_usd"])) - Decimal(str(month["base_cap_usd"]))
                    self._bonus = max(bonus, Decimal(0)) if bonus.is_finite() else None
                reading = month if self._value is not None else None
            except Exception:  # noqa: BLE001 - unreadable is unknown, never a number
                self._value = self._bonus = None
            if reading is not None and self.meter is not None:
                self._feed(reading)
            return self._value

    def _feed(self, month: dict[str, Any]) -> None:
        """One reading to OpenAI's meter. A reading the meter refuses leaves it unread (and OpenAI's
        reservations refused after 180 s); it never makes the month itself unreadable here."""
        try:
            # `cap`: the month's own line, which the House's OpenAI line never reads above while this
            # reading is fresh (`CampaignBudget.remaining`).
            self.meter.observe_month("openai", str(month.get("month") or ""), month["spent_usd"],
                                     settled=month.get("settled_usd"), cap=month["cap_usd"], previous=month.get("previous"),
                                     evidence={k: month.get(k) for k in ("cap_usd", "calls", "inflight_usd")})
        except Exception:  # noqa: BLE001 - see above
            pass

    def refresh(self) -> bool:
        """Read the month again if the last reading is older than the TTL. True when OpenAI's meter
        is fresh afterwards (or, with no meter, when the month could be read)."""
        self.remaining()
        return bool(self.meter.ready("openai")) if self.meter is not None else self._value is not None

    def profit_bonus(self) -> Decimal | None:
        """What the gateway's profit indexing adds to the month's cap, from the same reading as
        `remaining`: None when the gateway cannot be read or does not index."""
        self.remaining()
        return self._bonus


def frontier_tier(remaining: Decimal | None, reserve: Any) -> str:
    """Which frontier work the month still pays for, from `game.json` `frontier_reserve`.

    "all": everything. "earned": the unearned scheduled roles (operator, teacher, designer) and
    cheap-model research stop, so what is left goes to the roles that write code and to the work
    agents pay for. "audits": only audits and consultations agents buy with credits they earned.
    Unknown (None) is "all": the gateway still refuses at its line."""
    if remaining is None or not reserve:
        return "all"
    if remaining < Decimal(str(reserve.get("code_roles_usd", "0"))):
        return "audits"
    if remaining < Decimal(str(reserve.get("earned_usd", "0"))):
        return "earned"
    return "all"


#: The roles each tier still runs.
TIER_ROLES = {"all": None, "earned": frozenset({"architect", "toolsmith"}), "audits": frozenset()}

File: gateway/lib/frontier.mjs
// The frontier model, behind a budget (Sept 19, 2026).
//
// The floor's agents run on cheap open models. A frontier model audits a candidate before it
// trades real money and, on a slow clock, writes new strategy code. Its key lives here and
// nowhere else, for the same reason the venue keys do: the trading VM can ask, it cannot spend.
//
// The rule is a monthly dollar budget the VM cannot raise. A call is priced twice. Before it
// leaves, at its worst case (every input token uncached, every allowed output token used), and
// that much is reserved; a call whose worst case does not fit in what is left of the month is
// refused. After it returns, at what the provider says it used, and the difference is given
// back. A model this module has no price for is refused: an unpriced call is an uncapped one.

import { parseUsdMicro } from './money.mjs';

export const HOST = 'https://api.openai.com';
export const PATH = '/v1/responses';
export const AGENT_HEADER = 'X-LTCM-Agent';
//: The role a call is made for (a slug). It names the call's output ceiling in `FRONTIER_ROLE_MAX_OUTPUT`.
export const ROLE_HEADER = 'X-LTCM-Role';
//: The most output one call may ask for, unless its role names another ceiling. The reservation is sized from it.
export const MAX_OUTPUT_TOKENS = 16000;
//: The most any role's ceiling may name: a larger number in `FRONTIER_ROLE_MAX_OUTPUT` is not read.
export const ROLE_MAX_OUTPUT_LIMIT = 128000;
const SLUG = /^[a-z0-9][a-z0-9_-]{0,63}$/;
const RATE_FIELDS = ['input', 'cached', 'output', 'long_input', 'long_cached', 'long_output', 'uncached', 'long_uncached'];

/**
 * One row of rates, checked, or null. `base` is the standard row a flex row discounts (Sept 26, 2026 (the options-swarm
 * run, Wave 5)): a long-context rate the flex row leaves out is the standard one (the meter errs high), and no flex rate
 * may exceed its standard rate, so a flex call settled at flex rates never settles above the standard worst case it
 * reserved.
 */
function rates(row, base = null) {
  if (!row || typeof row !== 'object' || Array.isArray(row)) return null;
  const input = Number(row.input), cached = Number(row.cached ?? row.input), output = Number(row.output);
  const longInput = Number(row.long_input ?? base?.long_input ?? input), longCached = Number(row.long_cached ?? base?.long_cached ?? cached);
  const longOutput = Number(row.long_output ?? base?.long_output ?? output);
  // `input` is the cache-write rate, the dearest an input token can be. `uncached` is the
  // plain rate for a token neither read from nor written to the cache; absent, it is `input`,
  // so a table without it prices exactly as before.
  const uncached = Number(row.uncached ?? input), longUncached = Number(row.long_uncached ?? base?.long_uncached ?? longInput);
  if (![input, cached, output, longInput, longCached, longOutput, uncached, longUncached].every(Number.isFinite)
      || !(input > 0 && cached >= 0 && cached <= input && output > 0
        && longInput >= input && longCached >= cached && longCached <= longInput && longOutput >= output
        && uncached >= cached && uncached <= input && longUncached >= longCached && longUncached <= longInput && longUncached >= uncached)) {
    return null;
  }
  const out = { input, cached, output, long_input: longInput, long_cached: longCached, long_output: longOutput, uncached, long_uncached: longUncached };
  if (base && RATE_FIELDS.some(field => out[field] > base[field])) return null;
  return out;
}

/**
 * `{ model: { input, cached, output, ..., flex? } }` in dollars per million tokens, from `FRONTIER_MODELS`. A model whose
 * standard rates do not read is absent (refused). `flex` (Sept 26, 2026, Wave 5) is the model's flex-tier rates,
 * `{input, cached, output}` and optionally the long-context and uncached rates; a model without it, or whose flex rates
 * do not read, is priced as before and cannot be sent on the flex tier.
 */
export function priceTable(env = {}) {
  try {
    const table = JSON.parse(env.FRONTIER_MODELS || '{}');
    const out = {};
    for (const [model, row] of Object.entries(table)) {
      if (!/^[A-Za-z0-9._:-]{1,80}$/.test(model)) continue;
      const standard = rates(row);
      if (!standard) continue;
      const flex = row?.flex === undefined ? null : rates(row.flex, standard);
      out[model] = flex ? { ...standard, flex } : standard;
    }
    return out;
  } catch {
    return {};
  }
}

/** `FRONTIER_ROLE_MAX_OUTPUT` read: `{ role: ceiling }` for each slug with a whole number from 1 to the limit. */
export function roleCeilings(env = {}) {
  let table;
  try {
    table = JSON.parse(env.FRONTIER_ROLE_MAX_OUTPUT || '{}');
  } catch {
    return {};
  }
  if (!table || typeof table !== 'object' || Array.isArray(table)) return {};
  const out = {};
  for (const [role, value] of Object.entries(table)) {
    if (SLUG.test(role) && Number.isSafeInteger(value) && value >= 1 && value <= ROLE_MAX_OUTPUT_LIMIT) out[role] = value;
  }
  return out;
}

/**
 * The output ceiling of a call made for `role` (the `X-LTCM-Role` header): what `FRONTIER_ROLE_MAX_OUTPUT` names for it,
 * else MAX_OUTPUT_TOKENS. A role it does not name, a header that is not a slug, or none at all gets the default.
 */
export function outputCeiling(env = {}, role = null) {
  const table = roleCeilings(env);
  return typeof role === 'string' && SLUG.test(role) && Object.hasOwn(table, role) ? table[role] : MAX_OUTPUT_TOKENS;
}

/** A funded month expires instead of creating another allowance at the next UTC month boundary. */
export function monthCapMicro(env = {}, at = Date.now()) {
  if (Object.hasOwn(env, 'FRONTIER_FUNDED_MONTH')) {
    const funded = String(env.FRONTIER_FUNDED_MONTH);
    const time = new Date(at);
    if (!/^\d{4}-(0[1-9]|1[0-2])$/.test(funded) || !Number.isFinite(time.getTime())
        || time.toISOString().slice(0, 7) !== funded) return 0n;
  }
  return parseUsdMicro(env.FRONTIER_MONTH_USD, 0n);
}

const micro = dollars => BigInt(Math.ceil(dollars * 1e6));

/** The most this call can cost, in micro-dollars. */
export function worstCase(price, bodyBytes, maxOutputTokens) {
  // Text-only requests; a byte per possible token plus framing, without guessing a tokenizer.
  const inputTokens = bodyBytes + 4096;
  return micro((inputTokens * (price.long_input ?? price.input) + maxOutputTokens * (price.long_output ?? price.output)) / 1e6);
}

/** What the call did cost, from the provider's own usage block; null when it carries none. */
export function actualCost(price, usage) {
  if (!usage || typeof usage !== 'object') return null;
  const input = Number(usage.input_tokens), output = Number(usage.output_tokens);
  if (!Number.isFinite(input) || !Number.isFinite(output) || input < 0 || output < 0) return null;
  const cached = Math.min(input, Math.max(0, Number(usage.input_tokens_details?.cached_tokens) || 0));
  // GPT-5.6 and later report the tokens a call wrote to the prompt cache, billed at 1.25x the
  // plain input rate; reads are 0.1x (developers.openai.com/api/docs/guides/prompt-caching,
  // read Sept 22, 2026). A usage block without the field prices every uncached token as a write.
  const reported = Number(usage.input_tokens_details?.cache_write_tokens);
  const written = Number.isFinite(reported) && reported >= 0 ? Math.min(input - cached, reported) : input - cached;
  const plain = input - cached - written;
  const long = input > 272000;
  return micro((written * (long ? price.long_input ?? price.input : price.input)
    + plain * (long ? price.long_uncached ?? price.long_input ?? price.input : price.uncached ?? price.input)
    + cached * (long ? price.long_cached ?? price.cached : price.cached)
    + output * (long ? price.long_output ?? price.output : price.output)) / 1e6);
}

/**
 * True for the one 5xx that is settled at zero: the provider's own capacity refusal, a 503 whose
 * body is OpenAI's error object of type `service_unavailable_error` ("the requested model does not
 * have enough capacity to process your request at the moment", developers.openai.com/api/docs/
 * guides/error-codes, read Sept 24, 2026). The request was turned away before any work, so there is
 * nothing to bill. Measured Sept 22, 2026: two such answers ("server_is_overloaded") each kept its
 * whole worst case on the month, $2.02 for one.
 *
 * Every other 5xx keeps its worst case, because unknown is not free: a 500 `server_error` can
 * come after the model has worked, and the same documentation says nothing of whether it bills;
 * a 502, a 504 or a 503 without that body is an edge or proxy answering for a call the provider
 * may still have received and billed.
 */
export function unprocessed(status, text) {
  if (status !== 503) return false;
  try {
    const error = JSON.parse(text)?.error;
    return !!error && typeof error === 'object' && !Array.isArray(error) && error.type === 'service_unavailable_error';
  } catch {
    return false;
  }
}

/** The prompt-cache hints OpenAI documents for the Responses API, and nothing else. */
const CACHE_KEY = /^[A-Za-z0-9._:-]{1,64}$/;
const RETENTION = new Set(['in_memory', '24h']);

function cacheHintsValid(body) {
  if (body.prompt_cache_key !== undefined && !(typeof body.prompt_cache_key === 'string' && CACHE_KEY.test(body.prompt_cache_key))) return false;
  if (body.prompt_cache_retention !== undefined && !RETENTION.has(body.prompt_cache_retention)) return false;
  const options = body.prompt_cache_options;
  if (options !== undefined) {
    // A prewarm writes the cache without an answer; nothing here needs one, so it is refused
    // rather than trusted to be metered like an ordinary call.
    if (!options || typeof options !== 'object' || Array.isArray(options)
        || Object.keys(options).some(key => !['mode', 'ttl', 'prewarm'].includes(key))
        || (options.mode !== undefined && !['implicit', 'explicit'].includes(options.mode))
        || (options.ttl !== undefined && options.ttl !== '30m')
        || (options.prewarm !== undefined && options.prewarm !== false)) return false;
  }
  return true;
}

/** An input message's content: a string, or text blocks that may mark an explicit cache breakpoint. */
function textContent(content) {
  if (typeof content === 'string') return 0;
  if (!Array.isArray(content) || content.length < 1 || content.length > 16) return null;
  let breakpoints = 0;
  for (const block of content) {
    if (!block || typeof block !== 'object' || block.type !== 'input_text' || typeof block.text !== 'string'
        || Object.keys(block).some(key => !['type', 'text', 'prompt_cache_breakpoint'].includes(key))) return null;
    if (block.prompt_cache_breakpoint !== undefined) {
      const mark = block.prompt_cache_breakpoint;
      if (!mark || typeof mark !== 'object' || Array.isArray(mark) || Object.keys(mark).length !== 1 || mark.mode !== 'explicit') return null;
      breakpoints += 1;
    }
  }
  return breakpoints;
}

/**
 * Check a request body before it is sent. `{ model, price, flex, maxOutput, ceiling }` or `{ error, status }`.
 * Streaming and background calls are refused: the usage block that settles the bill arrives
 * with a complete response, and a call that outlives the request cannot be settled at all.
 * `flex` is the model's flex rates when the call asks for the flex tier, else null; the call is
 * still reserved at `price`, the standard rates, because OpenAI may serve it on another tier.
 */
export function admit(body, env, { role = null } = {}) {
  if (!body || typeof body !== 'object' || Array.isArray(body)) return { error: 'The request must be a JSON object.', status: 400 };
  const model = String(body.model || '');
  const price = priceTable(env)[model];
  if (!price) return { error: `No price is configured for model "${model}"; an unpriced call is refused.`, status: 403 };
  if (body.stream === true || body.background === true) return { error: 'Streaming and background calls cannot be metered and are refused.', status: 400 };
  // The meter prices only inline text and standard inference. A stored conversation, image,
  // paid built-in tool or faster service tier could bill work that is absent from this body.
  // Prompt-cache hints change what a call costs only through the usage block the bill is settled
  // from (cache reads are cheaper, writes are priced at `input`), so they are admitted, bounded.
  const allowed = new Set(['model', 'input', 'max_output_tokens', 'reasoning', 'stream', 'background', 'service_tier',
    'prompt_cache_key', 'prompt_cache_retention', 'prompt_cache_options']);
  let breakpoints = 0;
  const textInput = typeof body.input === 'string' || (Array.isArray(body.input) && body.input.every(item => {
    if (!item || typeof item !== 'object' || !['system', 'developer', 'user', 'assistant'].includes(item.role)
        || !Object.keys(item).every(key => ['role', 'content'].includes(key))) return false;
    // Assistant turns stay plain strings: output blocks are a different type, and none is needed.
    const marks = item.role === 'assistant' ? (typeof item.content === 'string' ? 0 : null) : textContent(item.content);
    if (marks === null) return false;
    breakpoints += marks;
    return true;
  }));
  // Flex (Sept 26, 2026 (the options-swarm run, Wave 5)): half price, slower, and refused with a 429 when OpenAI lacks
  // the capacity. Priority and every other tier stay refused: they bill above the standard rates this meter reserves.
  if (!textInput || Object.keys(body).some(key => !allowed.has(key))
      || (body.service_tier !== undefined && body.service_tier !== 'default' && body.service_tier !== 'flex')) {
    return { error: 'Only inline text on the standard or flex service tier is priced by this gateway.', status: 400 };
  }
  const flex = body.service_tier === 'flex' ? price.flex ?? null : null;
  if (body.service_tier === 'flex' && !flex) {
    return { error: `No flex price is configured for model "${model}"; a flex call is refused.`, status: 403 };
  }
  // OpenAI takes at most four cache writes a request.
  if (!cacheHintsValid(body) || breakpoints > 4) {
    return { error: 'Prompt-cache hints must be a bounded key, a documented retention, a 30m ttl without prewarm and at most four explicit breakpoints.', status: 400 };
  }
  // The ceiling is the role's (Sept 26, 2026, Wave 5: the weekly post-mortem writes more than 16,000 tokens), else
  // 16,000; the reservation is sized from the max_output_tokens admitted under it.
  const ceiling = outputCeiling(env, role);
  const maxOutput = Number(body.max_output_tokens);
  if (!Number.isInteger(maxOutput) || maxOutput < 1 || maxOutput > ceiling) {
    return { error: `max_output_tokens is required, between 1 and ${ceiling}.`, status: 400 };
  }
  return { model, price, flex, maxOutput, ceiling };
}
